espi: it8xxx2: enable configuration of Chromebook

Dino-Li opened the pull request:
This enables the below configuration so the AP and EC are able to
communicate over eSPI:

CONFIG_ESPI_PERIPHERAL_EC_HOST_CMD
CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION
CONFIG_ESPI_PERIPHERAL_CUSTOM_OPCODE

Signed-off-by: Dino Li <Dino.Li@ite.com.tw>

Dino-Li committed:
espi: it8xxx2: enable configuration of Chromebook

This enables the below configuration so the AP and EC are able to
communicate over eSPI:

CONFIG_ESPI_PERIPHERAL_EC_HOST_CMD
CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION
CONFIG_ESPI_PERIPHERAL_CUSTOM_OPCODE

Signed-off-by: Dino Li <Dino.Li@ite.com.tw>

keith-zephyr commented on drivers/espi/espi_it8xxx2.c:
```diff
@@ -153,6 +155,97 @@ static const struct ec2i_t pmc1_settings[] = {
+	defined(CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION)
+/* Host to RAM (H2RAM) memory mapping, 4K aligned */
+#define IT8XXX2_ESPI_H2RAM_POOL_SIZE 0x1000
+static uint8_t h2ram_pool[IT8XXX2_ESPI_H2RAM_POOL_SIZE] __aligned(0x1000);
```
The cros-ec driver allocated this pool in a separate linker section so it was always at a fixed address.   Allocating this into the .bss section could potentially waste up to 4 KiB of RAM due to the alignment requirement.

Dino-Li replied:
Putting this pool inside the first section of the RAM should be able to avoid gap. Can we fix this symptom later?  We need to add a location where is at the beginning of the RAM to zephyr_linker_sources().

brockus-zephyr replied:
I would prefer to see this gap avoided and not wait until later to fix the symptom

keith-zephyr replied:
RAM size is already low on the Hayato build (86% used).  So please address the alignment issue.

Dino-Li replied:
Done.
The first one commit is required to avoid the gap.

keith-zephyr commented on drivers/espi/espi_it8xxx2.c:
```diff
@@ -153,6 +155,97 @@ static const struct ec2i_t pmc1_settings[] = {
+		(CONFIG_ESPI_PERIPHERAL_HOST_CMD_PARAM_PORT_NUM >> 4) & 0xff;
+	/* Set host RAM window 0 size. (allow R/W) */
+	smfi_reg->SMFI_HRAMW0AAS =
+		find_msb_set(CONFIG_ESPI_IT8XXX2_HC_H2RAM_SIZE / 16) - 1;
```
Create a helper macro for setting the host RAM window size field.  The macro should also mask with 0x7 before updating the register.

Dino-Li replied:
Done.

keith-zephyr commented on drivers/espi/espi_it8xxx2.c:
```diff
@@ -153,6 +155,97 @@ static const struct ec2i_t pmc1_settings[] = {
+	smfi_reg->SMFI_HRAMWC |= (SMFI_H2RAMPS | SMFI_H2RAMW1E);
+#endif
+}
+#endif
```
Document which #if is closed here.  I think `defined(CONFIG_ESPI_PERIPHERAL_EC_HOST_CMD) || defined(CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION)`

Dino-Li replied:
Done.

Dino-Li committed:
riscv: linker: move snippets-ram-sections.ld to ram start

This align the RISC-V linker script with ARM linker script.
(The snippets-ram-sections.ld is included first inside the
RAMABLE_REGION)

Signed-off-by: Dino Li <Dino.Li@ite.com.tw>

tejlmand commented on drivers/espi/espi_it8xxx2.c:
```diff
@@ -153,6 +155,106 @@ static const struct ec2i_t pmc1_settings[] = {
+
+#if defined(CONFIG_ESPI_PERIPHERAL_EC_HOST_CMD) || \
+	defined(CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION)
+/* Host to RAM (H2RAM) memory mapping, 4K aligned */
+#define IT8XXX2_ESPI_H2RAM_POOL_SIZE 0x1000
+static uint8_t h2ram_pool[IT8XXX2_ESPI_H2RAM_POOL_SIZE]
+					__attribute__((section(".h2ram_pool")));
```
From where do this 4k alignment requirement originates ?
I think this code should have a comment describing why there is such a requirement.

Dino-Li replied:
Done.

tejlmand commented on drivers/espi/Kconfig.it8xxx2:
```diff
@@ -21,6 +21,29 @@ config ESPI_PERIPHERAL_HOST_IO
+	depends on ESPI_PERIPHERAL_ACPI_SHM_REGION
+	default 256
+	help
+	  The valid value for this option is 16/32/64/128/256/512/1024/2048 bytes.
```
Should this be a choice to make it clearer what values are valid ?

Dino-Li replied:
Done.

tejlmand reviewed (changes requested):
Thanks for this, some question mainly around the ld linker script template updates.

Dino-Li commented on include/arch/riscv/common/linker.ld:
```diff
@@ -208,6 +208,14 @@ SECTIONS
+/* Located in generated directory. This file is populated by calling
+ * zephyr_linker_sources(RAM_START ...).
+ */
+#include <snippets-ram-start.ld>
```
Done.
Let's align the riscv linker script with arm linker scripts.

Dino-Li commented on drivers/espi/Kconfig.it8xxx2:
```diff
@@ -21,6 +21,29 @@ config ESPI_PERIPHERAL_HOST_IO
+	depends on ESPI_PERIPHERAL_EC_HOST_CMD
+	default 256
+	help
+	  The valid value for this option is 16/32/64/128/256/512/1024/2048 bytes.
```
Done.

Dino-Li commented on soc/riscv/riscv-ite/it8xxx2/h2ram.ld:
```diff
@@ -0,0 +1,12 @@
+	_h2ram_pool_end = .;
+	_h2ram_pool_size = ABSOLUTE(_h2ram_pool_end - _h2ram_pool_start);
+	ASSERT(((ABSOLUTE(_h2ram_pool_start) & 0xfff) == 0),
+		"h2ram_pool section isn't 4k-byte aligned");
```
Since we have ensured the section is 4k-bytes aligned so let's remove the assertion.

keith-zephyr commented on drivers/espi/Kconfig.it8xxx2:
```diff
@@ -21,6 +21,81 @@ config ESPI_PERIPHERAL_HOST_IO
+config ESPI_PERIPHERAL_EC_HOST_CMD
+	default y
+
+choice
```
Add a  label to the choice.  This allows boards to override if needed.
```
choice IT8XXX2_H2RAM_HC_SIZE_CHOICE
```

Dino-Li replied:
Done.

keith-zephyr commented on drivers/espi/Kconfig.it8xxx2:
```diff
@@ -21,6 +21,81 @@ config ESPI_PERIPHERAL_HOST_IO
+config ESPI_PERIPHERAL_ACPI_SHM_REGION
+	default y
+
+choice
```
`choice IT8XXX2_H2RAM_ACPI_SHM_SIZE_CHOICE`

Dino-Li replied:
Done.

keith-zephyr commented on drivers/espi/espi_it8xxx2.c:
```diff
@@ -153,6 +155,112 @@ static const struct ec2i_t pmc1_settings[] = {
+	CONFIG_ESPI_IT8XXX2_HC_H2RAM_SIZE) > IT8XXX2_ESPI_H2RAM_POOL_SIZE)
+#error "EC host command parameters out of h2ram"
+#endif
+
```
This could also use checks to verify the ACPI and HC sections don't overlap.

Dino-Li replied:
Done.

Dino-Li commented on soc/riscv/riscv-ite/it8xxx2/CMakeLists.txt:
```diff
@@ -1,3 +1,8 @@
 	soc.c
 )
+
+zephyr_linker_sources_ifdef(CONFIG_ESPI_IT8XXX2
```
Done.

Dino-Li commented:
Rebase and fix conflict of linker.ld.

keith-zephyr commented:
@tejlmand Can you please re-review?

tejlmand commented on drivers/espi/Kconfig.it8xxx2:
```diff
@@ -21,6 +21,81 @@ config ESPI_PERIPHERAL_HOST_IO
+	default y
+
+choice IT8XXX2_H2RAM_HC_SIZE_CHOICE
+	prompt "choose a H2RAM space for ec host command"
```
Just:
```suggestion
	prompt "H2RAM space for ec host command"
```

as per https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#prompt-strings

and looks cleaner in `menuconfig` also.

Dino-Li replied:
Done.

tejlmand commented on drivers/espi/Kconfig.it8xxx2:
```diff
@@ -21,6 +21,81 @@ config ESPI_PERIPHERAL_HOST_IO
+choice IT8XXX2_H2RAM_HC_SIZE_CHOICE
+	prompt "choose a H2RAM space for ec host command"
+	depends on ESPI_PERIPHERAL_EC_HOST_CMD
+config IT8XXX2_H2RAM_HC_SIZE_256
```
You don't need to place `config IT8XXX2_H2RAM_HC_SIZE_256`, you can order the options ascending, and the specify the default as:
```
choice IT8XXX2_H2RAM_HC_SIZE_CHOICE
	prompt "H2RAM space for ec host command"
	default IT8XXX2_H2RAM_HC_SIZE_256
	depends on ESPI_PERIPHERAL_EC_HOST_CMD
```

Also add an empty line for readability.
```suggestion
 
config IT8XXX2_H2RAM_HC_SIZE_256
```

Dino-Li replied:
Done.

tejlmand commented on drivers/espi/Kconfig.it8xxx2:
```diff
@@ -21,6 +21,81 @@ config ESPI_PERIPHERAL_HOST_IO
+	depends on ESPI_PERIPHERAL_EC_HOST_CMD
+config IT8XXX2_H2RAM_HC_SIZE_256
+	bool "256"
+config IT8XXX2_H2RAM_HC_SIZE_16
```
please a blank line on top, also on the choices below.

Dino-Li replied:
Done.

tejlmand commented on drivers/espi/Kconfig.it8xxx2:
```diff
@@ -21,6 +21,81 @@ config ESPI_PERIPHERAL_HOST_IO
+	default y
+
+choice IT8XXX2_H2RAM_ACPI_SHM_SIZE_CHOICE
+	prompt "choose a H2RAM space for ACPI shared memory region"
```
Just
```suggestion
	prompt "H2RAM space for ACPI shared memory region"
```

Dino-Li replied:
Done.

tejlmand commented on drivers/espi/Kconfig.it8xxx2:
```diff
@@ -21,6 +21,81 @@ config ESPI_PERIPHERAL_HOST_IO
+choice IT8XXX2_H2RAM_ACPI_SHM_SIZE_CHOICE
+	prompt "choose a H2RAM space for ACPI shared memory region"
+	depends on ESPI_PERIPHERAL_ACPI_SHM_REGION
+config IT8XXX2_H2RAM_ACPI_SHM_SIZE_256
```
same comment regarding `default` and an empty line here and below.

Dino-Li replied:
Done.

tejlmand reviewed (commented):
looks good.
Some minor observations.

keith-zephyr commented:
@brockus-zephyr can you please review.

Dino-Li commented:
@tejlmand Could you please re-review?

tejlmand reviewed (dismissed):
lgtm

keith-zephyr commented:
@Dino-Li I believe if you rebase, it should fix the documentation failure.  I believe the doc failure was fixed by [PR#41865](https://github.com/zephyrproject-rtos/zephyr/pull/41865)

Dino-Li commented:
> @Dino-Li I believe if you rebase, it should fix the documentation failure. I believe the doc failure was fixed by [PR#41865](https://github.com/zephyrproject-rtos/zephyr/pull/41865)

@keith-zephyr Done.
